Refactor client error handling

JamesNK opened the pull request:
* Client throws OperationCanceledException when call is canceled using CancellationToken or times out with a deadline
* GetStatus should always work after a call failure
* Ensure call task is always completed when a call is completed or failed (previously some failure situations could leave it uncompleted)
* Provide status detail via call task
* Return headers from ResponseHeadersAsync when headers are available, even if response does not have valid gRPC headers

---

### OperationCanceledException vs RpcException

`OperationCanceledException` (OCE) is the intrinsic way to signal co-operative cancellation in .NET. There is a decent amount of code in .NET and async/await infrastructure that special cases OCE (and anything that derives from it). For example:
* Tasks with OCE thrown will have Canceled state
* OCE will not cause `TaskScheduler.UnobservedTaskException` event to be raised if no one waits for that task
* Built-in diagnostics and metrics for OCE

https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/task-cancellation

### Pending question:

What should happen for deadline. A deadline is basically a timeout, and after it is exceeded the client cancels the call. I think it makes sense that if you are awaiting the result from a gRPC call, and the deadline is exceeded, then you get an `OperationCanceledException`. The complication here is that the deadline is sent to the server, and there is a race between the client timer triggering the deadline cancellation (`OperationCanceledException`), and the server returning a deadline exceeded status (`RpcException`).

It would be weird to have two different exceptions thrown for the same thing, depending on who wins the race. Grpc.Core has this problem with the status on `RpcException`: https://github.com/grpc/grpc/issues/2685

Options to improve the situation:

1. Keep that status quo. Deadline cancellation is consistently RpcException
2. A deadline status returned from the server is thrown as OCE on the client. Deadline cancellation is consistently OCE.
3. It doesn't matter because theoretically the server should never win. The client starts its timer immediately. The server timer is delayed by the request and response time.

I like option 2: turn deadline status from the server into an OCE thrown on the client.

@jtattermusch @stephentoub @marklio

JamesNK commented on src/Grpc.Net.Client/Internal/GrpcCall.cs:
```diff
@@ -130,68 +126,55 @@ public void StartDuplexStreaming()
-                // 1. Its not exposed publically
-                // 2. Nothing else locks on call
-                // 3. We want to avoid allocating a private lock object
-                lock (this)
```
Nothing in dispose needs to be protected against being called multiple times.

JamesNK commented on src/Grpc.Net.Client/Internal/GrpcCall.cs:
```diff
@@ -202,163 +185,153 @@ public void EnsureNotDisposed()
+            // Clean up call resources once this call is finished
+            // Call may not be explicitly disposed when used with unary methods
+            // e.g. var reply = await client.SayHelloAsync(new HelloRequest());
+            DisposeCore(status.Value);
```
I have separated them so that there is a difference between "everything in this call is done, clean up all the resources" and "the user has explicitly called Dispose and now methods should throw `ObjectDisposedException`"

JamesNK commented on src/Grpc.Net.Client/Internal/GrpcCall.cs:
```diff
@@ -805,31 +788,25 @@ private static Status GetStatusCore(HttpResponseMessage httpResponseMessage)
+            // Call could have been canceled or deadline exceeded
+            if (_callCts.IsCancellationRequested)
+            {
+                throw new InvalidOperationException("Can't get the call trailers because the call was canceled.");
```
There is documentation on GetTrailers (the public API that uses this) that InvalidOperationException should be thrown if trailers aren't available.

JamesNK replied:
GetTrailers calls it.

JamesNK commented on src/Grpc.Net.Client/Internal/GrpcCall.cs:
```diff
@@ -202,165 +185,153 @@ public void EnsureNotDisposed()
-                    FinishResponse();
+                if (CallTask.IsCompletedSuccessfully)
+                {
+                    throw new RpcException(CallTask.Result);
```
If the `CallTask` is complete before a unary response message is read then something has gone wrong. Most likely the response headers were invalid for a gRPC call.

I'll add a comment.

JamesNK commented on src/Grpc.Net.Client/Internal/GrpcCall.cs:
```diff
@@ -202,165 +185,153 @@ public void EnsureNotDisposed()
-                ValidateTrailersAvailable();
+                if (_callTcs.Task.IsCompleted)
+                {
+                    return _callTcs.Task.Result;
```
Yes it should be IsCompletedSuccessfully. I fixed all of them yesterday but it must have gotten lost in a rebase.

jtattermusch commented:
IMHO there are two ways to think about how RPCs should behave:
- point of view 1) RPCs are just cancellable tasks like every other and therefore should comply with the framework guidelines for how such operations should behave. Intuitively, I don't feel that necessarily has to be the case, the RPCs are more complex entities than just simple "operations/tasks" and also because they come from the "distributed computing" world, they have some inherent complexity and modes of failure that makes them distinct from what I normally think of as a "Task". 
- point of view 2) RPCs are special entities that just happen to use CancellationToken as one of the inputs to provide a cancellation signal (=it's basically an implementation detail, other approach could have been used) and all the possible RPC errors are reported in a standard and uniform way.

Problems with the solution proposed in this PR:
- RPC errors are no longer reported uniformly. One now needs to have two catch blocks when invoking RPCs (RpcException and OperationCancelledException).
- for blocking unary calls, there is no way to access the call's status (because there is no call object). If no RpcException is thrown for the "Cancelled" case, the user also loses access to potentially important error details (error message and the trailing metadata which can carry extra details).
- this breaks the API compatibility between Grpc.Core and Grpc.Net, which is something we've been aiming for since the very start of grpc-dotnet. At the same time, Grpc.Core cannot really make this change because it would be a breaking change (and we just updated the major version in the previous release, now it's not a good time to do that again).
- because this would make Grpc.Core and Grpc.Net incompatible at the call invocation level, it would make it difficult for users to freely choose / switch between implementations (but we don't want users locked in to one of the two gRPC .NET implementations).


To me it seems that the problems above aren't easy ones to solve and we'd need more time (which we don't seem to have because of the upcoming release) to think through them properly so I'd rather not rush things and I'd prefer to stick with the "safe" option of keeping the existing API. Afterall the gRPC C# library (Grpc.Core) has been throwing RpcException("Cancelled") since 2015 and as far as I recall any users complaining about the API behavior being wrong with respect to cancellations - could this be more of a theoretical problem rather than something that will actually cause problems in real life use? (I think we can tolerate some degree of what will be perceived by some as less-than-ideal API in exchange for causing fewer disruptions to pre-existing user base that has already accommodated to the existing API with no noticeable friction)

jtattermusch commented:
a little comment on deadline handling:
From here https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclient.postasync?view=netframework-4.8 it looks like that 
`httpClient.PostAsync(uri, content, cancellationToken)`  throws  HttpRequestException in case of a timeout.

jtattermusch commented:
idea: maybe we could provide some opt-in syntactic sugar for cases where throwing OperationCancelledException is highly desired (I think in many cases just catching RpcException will be just fine). Something along the lines:
```
var call = client.FooBar();
// the extension method converts `call.ResponseAsync` to a Task that converts some RpcException errors into .NET frameworks exceptions (e.g. StatusCode.Cancelled becomes OperationCancelledException)
await call.ResponseWithCanonicalExceptionsAsync(); 
```

JamesNK commented:
> RPC errors are no longer reported uniformly. One now needs to have two catch blocks when invoking RPCs (RpcException and OperationCancelledException).

Only if someone is using cancellation, and only if they want to handle both situations.

> * for blocking unary calls, there is no way to access the call's status
> * this breaks the API compatibility between Grpc.Core and Grpc.Net, which is something we've been aiming for since the very start of grpc-dotnet
> * because this would make Grpc.Core and Grpc.Net incompatible at the call invocation level, it would make it difficult for users to freely choose / switch between implementations (but we don't want users locked in to one of the two gRPC .NET implementations).

Fair points. I think for 2 and 3 there are going to be some incompatibilities already because we haven't put a huge amount of effort into consistent exceptions, but I get that you don't want to make the situation worse.

I think throwing OperationCanceledException is [The Right Thing To Do](https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/exception-handling-task-parallel-library#exceptions-that-indicate-cooperative-cancellation). Extension methods aren't a good solution because OCE can come from any async method.

How about a setting on the channel to customize this behavior? `GrpcChannelOptions.ThrowCancellationRpcException`. By default grpc-dotnet would throw OCE, but if you want a consistent exception, or you want compatibility, then you could set the property to true when creating the channel.

JamesNK commented:
I've added a configuration switch to the PR. The change itself wasn't large, just deciding what exception to throw in a half dozen places. Most new code was replicating tests: https://github.com/grpc/grpc-dotnet/pull/479/commits/b4b7239d5bff102627a8a4fb983c1384373a6fca

jtattermusch commented:
The configuration switch at least gives users the opportunity to use behavior that's consistent with Grpc.Core (if they know about that switch and manually opt-in), but it's still problematic from the perspective of seamless migration between Grpc.Core and grpc-dotnet libraries. I assume vast majority of users will use the default option in each implementation and so we will end up with two incompatible user bases that have written a different code (one of them will have 2 catch blocks around each RPC and the other one just 1 ). 

One more problem is that the switch allows migration from Grpc.Core to grpc-dotnet (turn on the "legacy" switch), it doesn't allow migrating back (unless we add the same kind of option with an opposite default value to Grpc.Core) - and I'd argue if we want healthy ecosystem without lock-ins, users need to be able to flow both directions.

I understand that the introduction of OCE makes some some rough edges more smooth and solves some inconsistencies, but I still think overall this might bring more problems than what it solves. Because others seem to be of the opinion that throwing OCE is of great importance, perhaps we should gather more concrete examples of why it is such a big problem? (so far I've seen evidence that it's an inconsistency and it could cause some trouble, but no indication of how serious the trouble would be and how hard they would be to avoid - the counterindication is that in 4 years of Grpc.Core lifetime this hasn't come up so perhaps it's not that bad afterall)

jtattermusch commented:
Aside from the high-level debate on this PR, I've done some experiments regarding the "timeout race" and it seems that C-core now implements more robust logic for evaluating whether or not status code should be set as DeadlineExceeded on the client side and I wasn't able to reproduce grpc/grpc#2685 anymore. See https://github.com/grpc/grpc/pull/20158

JamesNK commented:
He is a concrete example of the problem:

```csharp
public async Task Export(Uri uri, CancellationToken cancellationToken = CancellationToken.None)
{
    // Throws OperationCanceledException
    cancellationToken.ThrowIfCancellationRequested();

    var httpClient = new HttpClient();

    // Throws OperationCanceledException
    var response = await httpClient.GetAsync(uri, cancellationToken);

    // Throws OperationCanceledException
    var token = await GetToken(response, cancellationToken);

    var channel = GrpcChannel.ForAddress(uri);
    var grpcClient = new MyApi.ApiClient(channel);

    // Throws RpcException <= INCONSISTENT with how other .NET frameworks report cancellation
    var data = await grpcClient.ExportAsync(CreateRequest(), cancellationToken);

    // Throws OperationCanceledException
    await File.WriteAsync("c:\data.csv", data.ToCsvBytes(), cancellationToken);
}
```

JamesNK commented:
Offline meeting discussion with @jtattermusch: He is ok with a setting if the default functionality is to throw `RpcException`, and it can be used to enable `OperationCanceledException`. Updated PR to match.

I've also added a note to the property that it is an experimental API, just in case we want to remove it or change how this is set in the future.
